src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
             {
-                // Snapshot the current set of buffered logs
-                // and set a new queue. This ensures that any new
-                // logs are written to the new buffer.
-                // We do this snapshot in a lock since Flush might be
-                // called by multiple threads concurrently, and we need
-                // to ensure we only log each log once.
-                currentBuffer = _logBuffer;
-                _logBuffer = new ConcurrentQueue<string>();
+                if (_flushTimer == null)
+                {
+                    // start a timer to flush accumulated logs in batches
+                    _flushTimer = new Timer(FlushCallback, null, LogFlushIntervalMs, LogFlushIntervalMs);
+                }
             }
+        }
 
-            if (currentBuffer.Count == 0)
+        private protected virtual void StopFlushTimer()
+        {
+            if (_flushTimer == null)
             {
                 return;
             }
 
-            // concatenate all lines into one string
-            StringBuilder sb = new StringBuilder();
-            string line = null;
-            while (currentBuffer.TryDequeue(out line))
+            lock (_flushTimerSyncRoot)
             {
-                sb.AppendLine(line);
+                if (_flushTimer != null)
+                {
+                    _flushTimer.Dispose();
+                    _flushTimer = null;
+                }
             }
+        }
 
-            if (_currentLogFileInfo == null)
-            {
-                // delay create log file
-                SetLogFile();
-            }
+        private void FlushCallback(object state)
+        {
+            Flush(true);
+        }
 
-            // write all lines in a single file operation
-            string content = sb.ToString();
-            try
-            {
-                AppendToFile(_currentLogFileInfo, content);
-            }
-            catch (Exception ex) when (ex is DirectoryNotFoundException || ex is FileNotFoundException)
+        public override void Flush()
+        {
+            Flush(false);
+        }
+
+        private protected virtual bool Flush(bool fromTimer)
+        {
+            if (_logBuffer.Count == 0)
             {
-                // need to handle cases where log files or directories might
-                // have been deleted from underneath us
-                SetLogFile();
-                AppendToFile(_currentLogFileInfo, content);
+                if (fromTimer)
+                {
+                    StopFlushTimer();
+
+                    // We just stopped our timer above; as a sanity check,
+                    // we ensure no traces were written to the buffer in the meantime.
+                    // If new traces showed up, we'll keep going and process them; otherwise,
+                    // we'll just return from here as new traces will re-enable our timer.
+                    if (_logBuffer.Count != 0)
```

Comment:
It isn't clear to me why you need this additional check and Start call. If the log buffer is not empty, it means Trace was called, and Trace starts the timer already, right?

Reply:
This is for the small window where a race would be possible:

Thread A | Thread B
--|--
Check buffer = 0 | Tracing, hasn't written to buffer yet
***** | Acquires timer lock and starts timer
Acquires timer lock and stops timer | -

It's unlikely this would ever happen. One thing we can do is flush from when the timer is stopped.